File: minimum_window_substring.py
class Solution:
    def minWindow(self, s: str, t: str) -> str:
        l = 0
        res = ""
        need, seen = {}, {}
        needcount, seencount = 0, 0

        for c in t:
            need[c] = need.get(c, 0) + 1
            needcount += 1

        for r in range(len(s)):
            seen[s[r]] = seen.get(s[r], 0) + 1

            if s[r] in need and seen[s[r]] <= need[s[r]]:
                seencount += 1

            while seencount >= needcount:
                window = s[l:r + 1]
                res = window if not res else min(res, window, key=len)
                seen[s[l]] -= 1
                if s[l] in need and seen[s[l]] < need[s[l]]:
                    seencount -= 1
                l += 1

        return res

File: minimumWindowSubstring.js
class Solution {
  /**
   * @param {string} s
   * @param {string} t
   * @return {string}
   */
  minWindow(s, t) {
    let l = 0;
    let seen = {},
      need = {};

    let res = "";

    let seencount = 0,
      needcount = 0;

    for (const c of t) {
      need[c] = 1 + (need[c] || 0);
      needcount++;
    }

    for (let r = 0; r < s.length; r++) {
      seen[s[r]] = 1 + (seen[s[r]] || 0);
      if (s[r] in need && seen[s[r]] <= need[s[r]]) {
        seencount++;
      }

      while (seencount >= needcount) {
        let window = s.slice(l, r + 1);
        res = !res || window.length < res.length ? window : res;
        seen[s[l]]--;
        if (s[l] in need && seen[s[l]] < need[s[l]]) {
          seencount--;
        }
        l++;
      }
    }
    return res;
  }
}
